sky/serve/autoscalers.py

```diff
@@ -132,93 +130,100 @@ def collect_request_information(
             if self.downscale_counter >= self.scale_down_consecutive_periods:
                 self.downscale_counter = 0
-                num_replicas_delta = target_num_replicas - num_replicas
+                return target_num_replicas
         else:
             self.upscale_counter = self.downscale_counter = 0
-        logger.info(f'Upscale counter: {self.upscale_counter}/'
-                    f'{self.scale_up_consecutive_periods}. '
-                    f'Downscale counter: {self.downscale_counter}/'
-                    f'{self.scale_down_consecutive_periods}')
+        return self.target_num_replicas
 
-        if num_replicas_delta == 0:
+    def evaluate_scaling(
+        self,
+        replica_infos: List['replica_managers.ReplicaInfo'],
+    ) -> List[AutoscalerDecision]:
+
+        # Don't count over-provision here.
+        self.target_num_replicas = self._get_desired_num_replicas()
+        logger.info(
+            f'Final target number of replicas: {self.target_num_replicas} '
+            f'({self.target_num_replicas} with '
+            f'over-provision), Upscale counter: {self.upscale_counter}/'
+            f'{self.scale_up_consecutive_periods}, '
+            f'Downscale counter: {self.downscale_counter}/'
+            f'{self.scale_down_consecutive_periods}')
+
+        alive_replica_infos = [info for info in replica_infos if info.is_alive]
+        num_alive_on_demand = len(alive_replica_infos)
+        logger.info(
+            f'Number of alive on-demand instances: {num_alive_on_demand}')
+
+        scaling_options = []
+        all_replica_ids_to_scale_down: List[int] = []
+
+        def _get_replica_ids_to_scale_down(
+            status_order: List['serve_state.ReplicaStatus'],
+            num_limit: int,
+        ) -> List[int]:
+            replica_ids_to_scale_down: List[int] = []
+            for target_status in status_order:
+                for info in alive_replica_infos:
+                    if info.status == target_status:
+                        if len(replica_ids_to_scale_down) >= num_limit:
+                            return replica_ids_to_scale_down
+                        replica_ids_to_scale_down.append(info.replica_id)
+            for info in alive_replica_infos:
+                if info.status not in status_order:
+                    if len(replica_ids_to_scale_down) >= num_limit:
+                        return replica_ids_to_scale_down
+                    replica_ids_to_scale_down.append(info.replica_id)
+            return replica_ids_to_scale_down
+
+        if num_alive_on_demand < self.target_num_replicas:
+            num_demand_to_scale_up = (self.target_num_replicas -
+                                      num_alive_on_demand)
+
+            for _ in range(num_demand_to_scale_up):
+                scaling_options.append(
+                    AutoscalerDecision(AutoscalerDecisionOperator.SCALE_UP,
+                                       target=num_demand_to_scale_up))
```

Comment:
number of replicas is consistent with the master branch.

Reply:
do you think we should change to resource override as in the master branch?

Reply:
Both of them sound great to me, but we should keep everything aligned - if we keep the "number of the replica to scale up" semantic, we should only add one `AutoscalerDecision`, instead of adding `num_demand_to_scale_up` times in a for loop.

Reply:
Also, the type of `AutoscalerDecision.target` should aligned with the semantic.

Reply:
updated